### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by pedrofaraco**
New version of the program, after a complete rewrite of it.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,31 +3,45 @@
...
+`#;CR_id;Name;EndDate;Type;isIRT?;isFEXERJ?`
+
+All players in the tournaments should already be in the first CSV.
+
+## How to run
 
 The user should run the following command:
 
-```python3 calculator.py```
+`python fexerj-rating-calculator.py <tournament_list_file> <first_torunament_to_run> <number_of_tournaments_to_run> <players_list_file>`
```
Comment (felipeamp) [inline README.md:24]: It feels like it's time to start using flags + argparse, because getting the order right is definitely not trivial :)

We could also leave that as a TODO

File: README.md [hunk 2]
```diff
@@ -3,31 +3,45 @@
...
+
+The program will create one intermediate players' list file for each tournament in the following format:
```
Comment (felipeamp) [inline README.md:26]: Will this be created in the same folder you run the code?

File: README.md [hunk 3]
```diff
@@ -3,31 +3,45 @@
...
+`RatingList_after_<number of the tournament>.csv`
+
+The program will also create a JSON file with all the players which ID was entered manually by the user. This is to avoid having to enter them again in a rerun. This file needs to be deleted by the user if the tournaments' file change.
```
Comment (felipeamp) [inline README.md:29]: all the players which ID -> all the players for which the ID

Comment (felipeamp) [inline README.md:29]: As a possible future improvement we could add a field to the json with the tournament file name and hash. This way we can automatically delete it if needed.

Could you file a feature request for this (and maybe other comments above):

File: README.md [hunk 4]
```diff
@@ -3,31 +3,45 @@
...
+
+
+## Command examples
+
+To run the first 25 tournaments of the file:
+
+```python3 fexerj-rating-calculator.py tournaments.csv 1 25 players.csv ```
```
Comment (felipeamp) [inline README.md:36]: Maybe we should document that it should start with index 1. Do we check at all that tournament indices aren't skipped/duplicated/out of order? I guess we should eventually add a few checks for this

File: README.md [hunk 5]
```diff
@@ -3,31 +3,45 @@
...
+
+To run only the 11th tournaments of the file:
 
-and answer the given questions. All players in the tournament should already be in one of the CSV's.
+```python3 fexerj-rating-calculator.py tournaments.csv 11 1 players.csv ```
 
 
 ## TODO
```
Comment (felipeamp) [inline README.md:43]: I guess many of my comments above fall in this category. Maybe just add them here. Ideally we should have issues enumerating these feature requests

File: classes.py [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+_CSV_DELIMITER = ';'
+_URLDOMAIN = "https://chess-results.com"
+_RATING_LIST_HEADER = 'Id_No;Id_CBX;Title;Name;Rtg_Nat;ClubName;Birthday;Sex;Fed;TotalNumGames;SumOpponRating' \
+                      ';TotalPoints'
+_MAX_NUM_GAMES_TEMP_RATING = 15
+_K_STARTING_NUM_GAMES = [(30, 0),  # grampo
+                         (25, _MAX_NUM_GAMES_TEMP_RATING),  # 15
+                         (15, 40),
+                         (10, 80)]
+_CURRENT_TOURNAMENT_SNR = 0
```
Comment (felipeamp) [inline classes.py:19]: What does SNR stand for?

File: classes.py [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+
+        self.tournaments_file = tournaments_file
+        self.first_item = first_item
+        self.items_to_process = items_to_process
+        self.rating_list = {}
+        self.tournaments = {}
+        self.cbx_to_fexerj = {}
+        self.manual_entries = {}
+
+        self.load_manual_entry_dict()
```
Comment (felipeamp) [inline classes.py:34]: Usually having a constructor that can fail and does non-trivial work is not great. I'd end the init in line 32 and make sure callers need to call a method called load_data or something similar (which open files etc, which might fail). This method would then do everything currently in the constructor from line 34 onwards

File: classes.py [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+                         (25, _MAX_NUM_GAMES_TEMP_RATING),  # 15
+                         (15, 40),
+                         (10, 80)]
+_CURRENT_TOURNAMENT_SNR = 0
+
+
+class FexerjRatingCycle:
+
+    def __init__(self, tournaments_file, first_item, items_to_process, initial_rating_filepath):
+
```
Comment (felipeamp) [inline classes.py:25]: I'm not a big fan of these extra lines at the beginning of files or in the middle of functions, unless they separate logically-distinct blocks (like in the middle of __init__ now).

File: classes.py [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+        self.cbx_to_fexerj = {}
+        self.manual_entries = {}
+
+        self.load_manual_entry_dict()
+
+        with open(tournaments_file, 'r') as f:
+            reader = csv.reader(f, delimiter=_CSV_DELIMITER)
+            self.tournaments = list(reader)[1:]
+            for tournament in self.tournaments:
+
```
Comment (felipeamp) [inline classes.py:40]: I'd also remove this extra newline

File: classes.py [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+                final_rating_filepath = "RatingList_after_%s.csv" % (tournament[0])
+
```
Comment (felipeamp) [inline classes.py:42]: and this one (and so on...)

File: classes.py [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+                if int(tournament[0]) in range(first_item, first_item + items_to_process):
+
+                    print("\nRunning tournament %s (%s)...\n" % (tournament[0], tournament[2]))
+
+                    self.get_rating_list(initial_rating_filepath)
+
+                    print("Reading from %s" % initial_rating_filepath)
+                    print("Writing to %s" % final_rating_filepath)
+
+                    trn_type = tournament[4]
```
Comment (felipeamp) [inline classes.py:52]: (opt) Nowadays Python3 supports enums. Maybe we should cast the string to an enum here.

File: classes.py [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+
+                    if trn_type == 'SS':
+                        tournament = SwissSingleTournament(self, tournament)
+                    elif trn_type == 'RR':
+                        tournament = RoundRobinTournament(self, tournament)
+                    elif trn_type == 'ST':
+                        tournament = SwissTeamTournament(self, tournament)
+                    else:
+                        raise ValueError('Wrong tournament type: %s' % trn_type)
```
Comment (felipeamp) [inline classes.py:61]: I'd also print the tournament #

File: classes.py [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+
+        self.write_manual_entry_dict()
+
+    def get_rating_list(self, initial_rating_filepath):
+
+        with open(initial_rating_filepath) as rating_list:
+            reader = csv.reader(rating_list, delimiter=_CSV_DELIMITER)
+            next(reader, None)  # Skip the headers
+            for row in reader:
+                self.rating_list.update({int(row[0]): FexerjPlayer(int(row[0]),  # ID FEXERJ
```
Comment (felipeamp) [inline classes.py:77]: Not sure if this is just in my screen, but could we reformat this in a way that the list stays a bit more to the left of the screen? For me they all show up as wrapped lines.

File: classes.py [hunk 14]
```diff
@@ -0,0 +1,544 @@
...
+                self.manual_entries = json.load(manual_entry_dict_file)
+
+    def write_manual_entry_dict(self):
+        with open('manual_entry_list.json', 'w') as manual_entry_dict_file:
+            manual_entry_dict_file.write(json.dumps(self.manual_entries))
+
+
+class FexerjPlayer:
+
+    def __init__(self, id_fexerj, id_cbx, title, name, lastrating, club, birthday, sex, federation, totgames,
```
Comment (felipeamp) [inline classes.py:104]: There should be a static factory method receiving a string and returning a FexerjPlayer. Currently we pass tons of args to the constructor, which is very error prone

This is not only an issue here, but above as well. We could also do smart things like

(self.id_fexerj, self.id_cbx, ...) = row
and cast to the right type afterwards

File: classes.py [hunk 15]
```diff
@@ -0,0 +1,544 @@
...
+                                                                   row[7],  # SEX
+                                                                   row[8],  # FEDERATION
+                                                                   int(row[9]),  # TOTAL NUM OF GAMES
+                                                                   row[10],  # SUM OF OPPONENT RATINGS
+                                                                   row[11])})  # POINTS AGAINST OPPONENTS
+                if len(row[1]) > 0:
+                    self.cbx_to_fexerj[int(row[1])] = int(row[0])
+
+    def load_manual_entry_dict(self):
+        if os.path.exists('manual_entry_list.json'):
```
Comment (felipeamp) [inline classes.py:93]: (opt) I'd add some logging that the manual_entry_list was found (or not). This might help debugging

File: classes.py (hunk 14 above)
Comment (felipeamp) [inline classes.py:104]: Could you split words with underscore where possible? lastrating -> last_rating etc.

It makes things a bit more readable

File: classes.py [hunk 16]
```diff
@@ -0,0 +1,544 @@
...
+        self.sex = sex
+        self.federation = federation
+        self.totGames = totgames
+        self.sumOppRating = sumopprating
+        self.ptsAgainstOpp = ptsagainstopp
+
+
+class TournamentPlayer:
+
+    def __init__(self, tournament, url):
```
Comment (felipeamp) [inline classes.py:122]: (nit) I'd rename url to player_url, to make it clearer what this is. Maybe add some pydoc string with a link to an example.

File: classes.py [hunk 17]
```diff
@@ -0,0 +1,544 @@
...
+        self.new_total_games = None
+        self.new_sum_oppon_ratings = None
+        self.new_pts_against_oppon = None
+
+    def load_player_page(self, url):
+
+        soup = BeautifulSoup(requests.get(url).content, 'lxml')
+        tables = soup.find_all("table", class_="CRs1")
+
+        if len(tables) == 0:
```
Comment (felipeamp) [inline classes.py:159]: `if not len(tables)` is a bit more idiomatic, believe it or not :)

File: classes.py [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+            del self
```
Comment (felipeamp) [inline classes.py:160]: Why do you need this? Won't the garbage collector handle this?

File: classes.py [hunk 19]
```diff
@@ -0,0 +1,544 @@
...
+            return
+
+        table = tables[0]
+        rows = table.find_all("tr")
+        for i in range(0, len(rows)):
```
Comment (felipeamp) [inline classes.py:165]: range(len(rows)) is a bit more idiomatic

File: classes.py [hunk 20]
```diff
@@ -0,0 +1,544 @@
...
+            cells = rows[i].find_all("td")
+            cell_data = [cell.get_text().strip() for cell in cells]
+            if cell_data[0] == "Name":
```
Comment (felipeamp) [inline classes.py:168]: (nit) I'd store cell_data[0] in a local varial before these if-elses

File: classes.py [hunk 21]
```diff
@@ -0,0 +1,544 @@
...
+                self.name = cell_data[1]
+            elif cell_data[0] == "Starting rank":
+                self.snr = cell_data[1]
+            elif cell_data[0] == "Ident-Number":
+                self.id = int(cell_data[1])
+                if not self.id:
```
Comment (felipeamp) [inline classes.py:174]: Is id 0 the default? I'm worried whether the cast above would fail and break everything

File: classes.py [hunk 22]
```diff
@@ -0,0 +1,544 @@
...
+                    manual_entry_key = str(self.tournament.ord) + '.' + str(self.snr)
```
Comment (felipeamp) [inline classes.py:175]: I think using "{}.{}".format() is cleaner and faster

File: classes.py [hunk 23]
```diff
@@ -0,0 +1,544 @@
...
+        for i in range(1, len(rows)):
+            cells = rows[i].find_all("td", recursive=False)
+            id_opponent = cells[id_cell_num].get_text().strip()
+            name_opponent = cells[name_cell_num].get_text().strip()
+            result_opponent = cells[result_cell_num].get_text().strip()
+            if name_opponent not in ['not paired', 'bye']:
+                self.add_opponent(int(id_opponent), name_opponent, result_opponent)
+
+    def add_opponent(self, sno, name, result):
+        if result[-1] != "K":
```
Comment (felipeamp) [inline classes.py:207]: again I'd store result[-1] in a local variable

File: classes.py [hunk 24]
```diff
@@ -0,0 +1,544 @@
...
+        self.new_total_games = self.last_total_games
+        self.new_sum_oppon_ratings = self.last_sum_oppon_ratings
+        self.new_pts_against_oppon = self.last_pts_against_oppon
+
+    def calculate_new_rating(self, is_fexerj_tournament):
+
+        invalid_opponents = []
+        for k, tp_oppon in self.opponents.items():
+            if tp_oppon[0].is_unrated:
+                if self.is_unrated or (tp_oppon[0].new_rating is None) or tp_oppon[0].new_rating == 0:
```
Comment (felipeamp) [inline classes.py:226]: could you add parenthesis around this if condition and break lines around the ORs?

You can also replace the last 2 conditions by 

not tp_oppon[0].new_rating

File: classes.py [hunk 25]
```diff
@@ -0,0 +1,544 @@
...
+                    invalid_opponents.append(k)
+        for i in invalid_opponents:
+            del self.opponents[i]
```
Comment (felipeamp) [inline classes.py:229]: Why is this needed?

File: classes.py [hunk 26]
```diff
@@ -0,0 +1,544 @@
...
+
+        self.this_games = len(self.opponents)
+
+        if self.this_games == 0 or (self.is_unrated and self.this_pts_against_oppon == 0):
```
Comment (felipeamp) [inline classes.py:233]: Again please break lines around ORs and add extra parenthesis around the condition if needed

File: classes.py [hunk 27]
```diff
@@ -0,0 +1,544 @@
...
+            self.opponents.append([sno, name, float(res)])
+
+    def keep_current_rating(self):
+        # self.new_rating = self.this_rating = self.last_rating
+        self.new_rating = self.last_rating
+        self.new_total_games = self.last_total_games
+        self.new_sum_oppon_ratings = self.last_sum_oppon_ratings
+        self.new_pts_against_oppon = self.last_pts_against_oppon
+
+    def calculate_new_rating(self, is_fexerj_tournament):
```
Comment (felipeamp) [inline classes.py:221]: This is quite a bit function. Could you add comments explaining the general high-level steps of what the function is doing?

File: classes.py [hunk 28]
```diff
@@ -0,0 +1,544 @@
...
+        self.this_games = len(self.opponents)
+
+        if self.this_games == 0 or (self.is_unrated and self.this_pts_against_oppon == 0):
+            self.keep_current_rating()
+            return
+
+        self.this_sum_oppon_ratings = 0
+        self.this_pts_against_oppon = 0
+        for snr_opp, oppon in self.opponents.items():
+            if oppon[0].is_unrated:  # unrated x unrated will never happen
```
Comment (felipeamp) [inline classes.py:240]: Why will it never happen?

File: classes.py [hunk 29]
```diff
@@ -0,0 +1,544 @@
...
+        self.last_k = self.get_current_k()
+        self.this_avg_oppon_rating = self.this_sum_oppon_ratings / self.this_games
+        rating_diff = self.this_avg_oppon_rating - self.last_rating
+        self.this_expected_points = self.this_games / (1.0 + 10.0 ** (rating_diff / 400.0))
+        self.this_points_above_expected = (self.this_pts_against_oppon - self.this_expected_points)
+
+        self.new_total_games = self.last_total_games + self.this_games
+
+        calc_rule = self.get_calculation_rule(is_fexerj_tournament)
+        if calc_rule == "TEMPORARY":
```
Comment (felipeamp) [inline classes.py:261]: We should also use enums for the calculation rule, and ideally have separate functions for each one of them (instead of having the entire logic in this function in each if-else branch)

File: classes.py [hunk 30]
```diff
@@ -0,0 +1,544 @@
...
+        if score == 1.0:
+            score = (num_valid_games + 0.5) / (num_valid_games + 1.0)
+        elif score == 0.0:
+            score = 0.5 / (num_valid_games + 1.0)
+        return avg_oppon_rating + 400.0 * math.log10(score / (1.0 - score))
+
+
+class Tournament:
+
+    def __init__(self, rating_cycle, tournament):
```
Comment (felipeamp) [inline classes.py:345]: I really like how you pass the tournament argument and split below. Ideally we should do the same for rows above (see other comment above).

File: classes.py [hunk 31]
```diff
@@ -0,0 +1,544 @@
...
+        self.is_irt = int(tournament[5])
+        self.is_fexerj = int(tournament[6])
+        self.players = {}
+        self.unrated_keys = []
+        self.temp_keys = []
+        self.established_keys = []
+        self.rating_cycle = rating_cycle
+        self.load_player_list()
+
+    def load_player_list(self):
```
Comment (felipeamp) [inline classes.py:360]: Why is this here (and called above)? Could we just delete it?

File: classes.py [hunk 32]
```diff
@@ -0,0 +1,544 @@
...
+                fp = self.rating_cycle.rating_list[self.rating_cycle.cbx_to_fexerj[tp.id]]
+            else:
+                fp = self.rating_cycle.rating_list[tp.id]
+
+            tp.last_rating = int(fp.lastRating)
+            tp.last_total_games = int(fp.totGames)
+            tp.last_sum_oppon_ratings = int(fp.sumOppRating)
+            tp.last_pts_against_oppon = float(fp.ptsAgainstOpp)
+
+            if int(fp.totGames) == 0:
```
Comment (felipeamp) [inline classes.py:375]: fp.totGames should either always be an int everywhere, or this cast should be done only once and stored in a local variable

**Review by felipeamp — CHANGES_REQUESTED** [review]
In general the usage of python looks good. In some places I find it a bit hard to follow because the function is either too large or the class has too many fields. I think this is also a criticism of the original code though. Ideally we should use more enums, have more classes and compose them where needed, and extract some things into new functions.

I'm usually nitpicky in my code reviews, so take my comments with a grain of salt.

File: README.md (hunk 3 above)
Comment (pedrofaraco) [reply README.md:29]: Fixed.

File: classes.py (hunk 8 above)
Comment (pedrofaraco) [reply classes.py:25]: Fixed.

File: classes.py (hunk 9 above)
Comment (pedrofaraco) [reply classes.py:40]: Fixed.

File: classes.py (hunk 10 above)
Comment (pedrofaraco) [reply classes.py:42]: Fixed.

File: classes.py (hunk 13 above)
Comment (pedrofaraco) [reply classes.py:77]: The method call has many parameters so a new line was added after each parameter to improve readability.

image

Could you please explain better how exactly do you want this to be presented?

File: classes.py (hunk 17 above)
Comment (pedrofaraco) [reply classes.py:159]: I don't. :-) But ok. Fixed.

File: classes.py (hunk 19 above)
Comment (pedrofaraco) [reply classes.py:165]: Ok, HERE i do believe is more idiomatic. :-) Fixed.

File: README.md [hunk 33]
```diff
@@ -3,31 +3,45 @@
...
-Id_No;Title;Name;Rtg_Nat;ClubName;Birthday;Sex;Fed;TotalNumGames
+There are two CSV needed: one for players and another for tournaments to be processed. The separator should be a semicolon ";". The first CSV should have the following header/columns:
 
-while the second CSV should have:
+`Id_No;Title;Name;Rtg_Nat;ClubName;Birthday;Sex;Fed;TotalNumGames;AvgOpponRating;TotalPoints`
 
-Id_No;Title;Name;Rtg_Nat;ClubName;Birthday;Sex;Fed;TotalNumGames;AvgOpponRating;TotalPoints
+While the second CSV should have:
 
+`#;CR_id;Name;EndDate;Type;isIRT?;isFEXERJ?`
```
[the remark this answers is not in the document]
Comment (pedrofaraco) [reply README.md:16]: Make sense. Do you mind to open an issue for that?

File: README.md (hunk 2 above)
Comment (pedrofaraco) [reply README.md:26]: Currently, yes. Should we generate the files in a subfolder like "NewRatingCycle"?

File: README.md (hunk 3 above)
Comment (pedrofaraco) [reply README.md]: You mean the JSON file name to be something like "Tournament_20_ManualEntries.json"?

File: README.md (hunk 4 above)
Comment (pedrofaraco) [reply README.md]: It does not start necessarily with index 1. For example, if we find some issue with a given tournament in Chess Results page, we fix it and rerun from that point of the cycle.

We can do a lot of checks on the input files. Maybe infinite ones. :-) We just need to determine which ones are worth implementing.

File: README.md (hunk 5 above)
Comment (pedrofaraco) [reply README.md:43]: Sure, I will create the issues.

File: README.md (hunk 1 above)
Comment (pedrofaraco) [reply README.md:24]: Opened as Issue #6.

File: classes.py (hunk 12 above)
Comment (pedrofaraco) [reply classes.py:52]: Opened as Issue #7.

File: classes.py (hunk 11 above)
Comment (pedrofaraco) [reply classes.py:44]: Opened as issue #8.

File: classes.py (hunk 15 above)
Comment (pedrofaraco) [reply classes.py:81]: Opened issue #9.

File: classes.py (hunk 14 above)
Comment (pedrofaraco) [reply classes.py:91]: Opened issue #10.

File: classes.py (hunk 16 above)
Comment (pedrofaraco) [reply classes.py:108]: Will open an issue for that. Can you please elaborate what kind of example you would expect on this specific case?

File: classes.py (hunk 18 above)
Comment (pedrofaraco) [reply classes.py:139]: Honestly, I don't recall why it was done this way. I would guess that it was a way to handle the occurrence of empty player rows on ChessResults page. So if by any means the soup.find_all returns an empty row, the TournamentPlayer object is destroyed. Any suggestion about how to implement this?

File: classes.py (hunk 20 above)
Comment (pedrofaraco) [reply classes.py:146]: Opened issue #11.

File: classes.py (hunk 21 above)
Comment (pedrofaraco) [reply classes.py:152]: If the player has no Fexerj ID on Chess Results (i.e. when the arbiter forgets to add it), line 173 will zero self.id variable. In that case, the program will enter the IF.

This is similar to the following  part of the master branch (original program):
image

Please let me know if you any specific change on this implementation.

File: classes.py (hunk 22 above)
Comment (pedrofaraco) [reply classes.py:153]: Opened issue #12.

File: classes.py (hunk 28 above)
Comment (pedrofaraco) [reply classes.py:212]: Because in the beginning of this method, if the player is unrated his unrated opponent will be moved to the invalid_opponents list. When the program reaches the part you pointed out, self.opponents will never have a unrated if the current player is also unrated.

File: classes.py (hunk 29 above)
Comment (pedrofaraco) [reply classes.py:229]: Opened issue #13

File: classes.py (hunk 30 above)
Comment (pedrofaraco) [reply classes.py:313]: Please clarify which specific comment you are referring to.

File: classes.py (hunk 31 above)
Comment (pedrofaraco) [reply classes.py:328]: Tournament is a parent class of other 3 child classes. The idea of line 358 was o make sure that load_player_list method would be called always called when creating object of the 3 child classes.

Would you have implemented this in a different way? Let me know.

File: classes.py (hunk 32 above)
Comment (pedrofaraco) [reply classes.py:341]: Opened issue #14.

File: classes.py (hunk 6 above)
Comment (pedrofaraco) [reply classes.py:19]: Chess Results uses this acronym for sequence numbers. I've used the same acronym on other parts of the code but this specific variable is not being used anywhere. Opened issue #15.

File: classes.py (hunk 7 above)
Comment (pedrofaraco) [reply classes.py:32]: Interesting. So you would never have a method call in a constructor?

File: classes.py (hunk 14 above)
Comment (pedrofaraco) [reply classes.py:91]: Will need your help on that... :-)

File: classes.py (hunk 23 above)
Comment (pedrofaraco) [reply classes.py:183]: Opened issue #16.

File: classes.py (hunk 24 above)
Comment (pedrofaraco) [reply classes.py:201]: Opened issue #17.

File: classes.py (hunk 25 above)
Comment (pedrofaraco) [reply classes.py:204]: Explained here: https://github.com/felipeamp/fexerj-rating-calculator/pull/1#discussion_r1900464892

File: classes.py (hunk 26 above)
Comment (pedrofaraco) [reply classes.py:206]: Added to issue #17.

File: classes.py (hunk 27 above)
Comment (pedrofaraco) [reply classes.py:197]: Added to issue #17.

File: README.md (hunk 2 above)
Comment (felipeamp) [reply README.md:26]: I think either is fine, as long as there is no filename collision

File: README.md (hunk 3 above)
Comment (felipeamp) [reply README.md]: Not necessarily, it could be a field in the json file. My suggestion was that if we kept the tournament name and a hash (to avoid name collisions) in the json file we could automatically delete json files that aren't needed. Anyway, this is an idea of a nice-to-have feature, feel free to ignore or file a feature request issue.

File: classes.py (hunk 28 above)
Comment (felipeamp) [reply classes.py:212]: Could you add this to the file comment please?

File: classes.py (hunk 7 above)
Comment (felipeamp) [reply classes.py:32]: You can, but ideally they should never fail (otherwise the caller would be an object in a broken state, and throwing exceptions from constructors is not great either).

So ideally the constructor should set the initial values to something that makes sense, and then the user could initialize them by calling the `load_manual_entry_dict` method.

File: README.md (hunk 4 above)
Comment (felipeamp) [reply README.md]: That makes sense. Could you add a file comment explaining this?

File: classes.py (hunk 13 above)
Comment (felipeamp) [reply classes.py:65]: yeah, this is not great, but I'm not sure of a great way of doing it. Keyword arguments would have pretty much the same issue. Maybe if you extracted the part that constructs the `FexerjPlayer` into a local variable outside of the method call the indentation would be smaller. Something like this:

```suggestion
for row in reader:
    player = FexerjPlayer(int(row[0]),  # ID FEXERJ
                                           row[1],  # ID CBX
                                           row[2],  # TITLE
                                           row[3],  # NAME
                                           int(row[4]),  # RATING
                                           row[5],  # CLUB
                                           row[6],  # BIRTHDAY
                                           row[7],  # SEX
                                           row[8],  # FEDERATION
                                           int(row[9]),  # TOTAL NUM OF GAMES
                                           row[10],  # SUM OF OPPONENT RATINGS
                                           row[11])  # POINTS AGAINST OPPONENTS
    self.rating_list.update({int(row[0]): player})
```

File: classes.py (hunk 18 above)
Comment (felipeamp) [reply classes.py:139]: I think you can just remove this line and let it return. You could also add a boolean return type informing whether it found the player or returned early, but personally I think it's not needed.

The issue with deleting self is that a caller could try to read from it and use a dangling pointer. If you just let python garbage collect this, all should be fine.

File: classes.py (hunk 21 above)
Comment (felipeamp) [reply classes.py:152]: This looks good to me, thanks for clarifying

File: classes.py (hunk 30 above)
Comment (felipeamp) [reply classes.py:313]: I suspect I meant the constructor of the `FexerjPlayer` and other classes, which have many arguments. But TBH it's been too long and I don't know for sure.

File: classes.py (hunk 31 above)
Comment (felipeamp) [reply classes.py:328]: Ah, I see what you mean. As I mentioned in another comment thread, I'd leave it for the caller to call `load_player_list`, and let the constructor just initialize the object fields to some sane values. If we do that, then we don't need this function in the `Tournament` class. 

I think with python 3.8 we can alternatively make the `Tournament` class inherit from `typing.Prototype`, to mark this method as "virtual" (https://docs.python.org/3/library/typing.html#typing.Protocol). I don't think we need to do this, but if you wanted to make sure the child classes have such a method, then this would be the cleanest way.

**Comment by felipeamp** [comment]
Thanks for the PR! I closed some comments and tried to clarify others. Let me know in case something still isn't clear!

File: classes.py (hunk 16 above)
Comment (felipeamp) [reply classes.py:108]: I mean an example of a valid URL, linking to a player in chess-results. Just so that callers know what URL format to pass

File: classes.py (hunk 14 above)
Comment (felipeamp) [reply classes.py:91]: Somehow my comment dissappeared. Ideally such a class should have either a factory or, even better, a builder class, so that you could construct it fluently (e.g. https://dzone.com/articles/fluent-builder-pattern). For now I think we can leave as is, since it's just a nice-to-have

File: README.md (hunk 4 above)
Comment (pedrofaraco) [reply README.md]: I extended line 38 of README.

File: classes.py (hunk 7 above)
Comment (pedrofaraco) [reply classes.py:32]: FexerjRatingCycle works as a static class, being called from the main (fexerj-rating-calculator.py), so there is not object instantiation. Your suggestion is to remove this method call from here and call it from the fexerj-rating-calculator.py file?

File: classes.py (hunk 13 above)
Comment (pedrofaraco) [reply classes.py:65]: Fixed

File: classes.py (hunk 7 above)
Comment (felipeamp) [reply classes.py:32]: Yes, that was my suggestion.

File: classes.py (hunk 16 above)
Comment (pedrofaraco) [reply classes.py:108]: Added a comment to the constructor.

File: classes.py (hunk 7 above)
Comment (pedrofaraco) [reply classes.py:32]: Fixed.

File: classes.py (hunk 28 above)
Comment (pedrofaraco) [reply classes.py:212]: Fixed.

Comment (pedrofaraco) [reply classes.py:212]: Fixed.

File: classes.py (hunk 31 above)
Comment (pedrofaraco) [reply classes.py:328]: Moved method call to callers and removed empty method from parent class.

File: README.md (hunk 3 above)
Comment (pedrofaraco) [reply README.md]: Opened issue #19 for that.

File: README.md (hunk 2 above)
Comment (pedrofaraco) [reply README.md:26]: Opened issue #20.

File: README.md (hunk 3 above)
Comment (pedrofaraco) [reply README.md]: @felipeamp, quick question about your suggestion of adding the tournament name and hash to the JSON file. I want to make sure I understand what you meant before implementing it.

I see three possible interpretations:
1. Hash of the tournament filename - store the tournaments filename (e.g. torneios_2024R2.csv) inside the JSON, plus a hash of that name to avoid collisions in case two different tournament files share the same name.
2. Hash of the entire file contents - store a hash of the tournaments CSV contents so the program can detect when the file has changed and automatically discard the JSON.
3. Both - store the filename and a hash of its contents for full identification and change detection.

If it's 2 or 3, there's also a question about what to do when the file changes. If we auto-discard the JSON, the user loses all their manual entries, which is risky if they just appended new rows. So should we auto-discard, or just warn the user and let them delete the JSON manually?

Happy to discuss.

Comment (felipeamp) [reply README.md]: My suggestion was to have a hash of the entire file contents, so that we could easily detect if anything has changed. 

Regarding the manual entries, isn't it possible that new users will have to be entered manually as well, is the tournament file changed? Maybe we should not merely discard the original one, but offer to merge against newly entered players.

I'm not sure what would be the best user interface for this, but maybe prompting each previously entry and ask whether to keep, delete or edit. Then they could enter any new players after this part is done. 

For a first step warning might be enough, and let the user manually edit things.
